File: server/internal/layers/presentation/models/track.go
package view_models

import (
	"context"
	"time"

	"github.com/gungun974/Melodink/server/internal/helpers"
	"github.com/gungun974/Melodink/server/internal/layers/domain/entities"
)

type TrackViewModel struct {
	Id int `json:"id"`

	UserId *int `json:"user_id"`

	Title    string `json:"title"`
	Duration int    `json:"duration"`

	TagsFormat string `json:"tags_format"`
	FileType   string `json:"file_type"`

	FileSignature  string `json:"file_signature"`
	CoverSignature string `json:"cover_signature"`

	Albums  []int `json:"albums"`
	Artists []int `json:"artists"`

	TrackNumber int `json:"track_number"`
	DiscNumber  int `json:"disc_number"`

	Metadata TrackMetadataViewModel `json:"metadata"`

	SampleRate       int  `json:"sample_rate"`
	BitRate          *int `json:"bit_rate"`
	BitsPerRawSample *int `json:"bits_per_raw_sample"`

	Score float64 `json:"score"`

	DateAdded string `json:"date_added"`
}

type TrackMetadataViewModel struct {
	TotalTracks int `json:"total_tracks"`
	TotalDiscs  int `json:"total_discs"`

	Date string `json:"date"`
	Year int    `json:"year"`

	Genres  []string `json:"genres"`
	Lyrics  string   `json:"lyrics"`
	Comment string   `json:"comment"`

	AcoustID string `json:"acoust_id"`

	MusicBrainzReleaseId   string `json:"music_brainz_release_id"`
	MusicBrainzTrackId     string `json:"music_brainz_track_id"`
	MusicBrainzRecordingId string `json:"music_brainz_recording_id"`

	Composer string `json:"composer"`
}

func ConvertToTrackViewModels(
	ctx context.Context,
	tracks []entities.Track,
) []TrackViewModel {
	tracksViewModels := make([]TrackViewModel, len(tracks))

	for i, track := range tracks {
		tracksViewModels[i] = ConvertToTrackViewModel(ctx, track)
	}

	return tracksViewModels
}

func ConvertToTrackViewModel(
	ctx context.Context,
	track entities.Track,
) TrackViewModel {
	albums := make([]int, len(track.Albums))

	for i, album := range track.Albums {
		albums[i] = album.Id
	}

	artists := make([]int, len(track.Artists))

	for i, artist := range track.Artists {
		artists[i] = artist.Id
	}

	metadata := track.Metadata

	if metadata.Genres == nil {
		metadata.Genres = []string{}
	}

	return TrackViewModel{
		Id: track.Id,

		UserId: track.UserId,

		Title:    track.Title,
		Duration: track.Duration,

		TagsFormat: track.TagsFormat,
		FileType:   track.FileType,

		FileSignature:  track.FileSignature,
		CoverSignature: track.CoverSignature,

		DateAdded: track.DateAdded.UTC().Format(time.RFC3339),

		Albums:  albums,
		Artists: artists,

		TrackNumber: metadata.TrackNumber,
		DiscNumber:  metadata.DiscNumber,

		Metadata: TrackMetadataViewModel{
			TotalTracks: metadata.TotalTracks,
			TotalDiscs:  metadata.TotalDiscs,

			Date: metadata.Date,
			Year: metadata.Year,

			Genres:  metadata.Genres,
			Lyrics:  metadata.Lyrics,
			Comment: metadata.Comment,

			AcoustID: metadata.AcoustID,

			MusicBrainzReleaseId:   metadata.MusicBrainzReleaseId,
			MusicBrainzTrackId:     metadata.MusicBrainzTrackId,
			MusicBrainzRecordingId: metadata.MusicBrainzRecordingId,

			Composer: metadata.Composer,
		},

		SampleRate:       track.SampleRate,
		BitRate:          track.BitRate,
		BitsPerRawSample: track.BitsPerRawSample,

		Score: getTrackScore(ctx, track),
	}
}

type TrackArtistRoleViewModel struct {
	Type string `json:"type"`

	Artist string `json:"artist"`

	Attributes []string `json:"attributes"`
}

func ConvertToTrackArtistsRolesViewModel(
	artistsRoles []entities.TrackArtistRole,
) []TrackArtistRoleViewModel {
	artistsRolesViewModels := make([]TrackArtistRoleViewModel, len(artistsRoles))

	for i, artistRole := range artistsRoles {
		artistsRolesViewModels[i] = ConvertToTrackArtistRoleViewModel(artistRole)
	}

	return artistsRolesViewModels
}

func ConvertToTrackArtistRoleViewModel(
	artistRole entities.TrackArtistRole,
) TrackArtistRoleViewModel {
	return TrackArtistRoleViewModel{
		Type: artistRole.Type,

		Artist: artistRole.Artist,

		Attributes: artistRole.Attributes,
	}
}

func getTrackScore(
	ctx context.Context,
	track entities.Track,
) float64 {
	if len(track.Scores) != 0 {
		user, err := helpers.ExtractCurrentLoggedUser(ctx)
		if err == nil {
			for _, trackScore := range track.Scores {
				if trackScore.UserId == user.Id {
					return trackScore.Score
				}
			}
		}
	}

	return 0.0
}

File: client/lib/features/track/domain/entities/track.dart
import 'package:equatable/equatable.dart';
import 'package:melodink_client/core/api/api.dart';
import 'package:melodink_client/core/helpers/double_to_string_without_zero.dart';
import 'package:melodink_client/features/library/domain/entities/album.dart';
import 'package:melodink_client/features/library/domain/entities/artist.dart';
import 'package:melodink_client/features/settings/domain/entities/settings.dart';
import 'package:melodink_client/features/track/domain/entities/track_compressed_cover_quality.dart';
import 'package:melodink_client/features/tracker/domain/entities/track_history_info.dart';

class Track extends Equatable {
  final int id;

  final String title;
  final Duration duration;

  final String tagsFormat;
  final String fileType;

  final String fileSignature;
  final String coverSignature;

  final List<Album> albums;
  final List<Artist> artists;

  final int trackNumber;
  final int discNumber;

  final TrackMetadata metadata;

  final int sampleRate;
  final int? bitRate;
  final int? bitsPerRawSample;

  final DateTime dateAdded;

  final double score;

  final TrackHistoryInfo? historyInfo;

  const Track({
    required this.id,
    required this.title,
    required this.duration,
    required this.tagsFormat,
    required this.fileType,
    required this.fileSignature,
    required this.coverSignature,
    required this.albums,
    required this.artists,
    required this.trackNumber,
    required this.discNumber,
    required this.metadata,
    required this.sampleRate,
    required this.bitRate,
    required this.bitsPerRawSample,
    required this.dateAdded,
    required this.score,
    this.historyInfo,
  });

  Track copyWith({
    int? id,
    String? title,
    Duration? duration,
    String? tagsFormat,
    String? fileType,
    String? fileSignature,
    String? coverSignature,
    List<Album>? albums,
    List<Artist>? artists,
    int? trackNumber,
    int? discNumber,
    TrackMetadata? metadata,
    int? sampleRate,
    int? bitRate,
    int? bitsPerRawSample,
    DateTime? dateAdded,
    double? score,
    TrackHistoryInfo? Function()? historyInfo,
  }) {
    return Track(
      id: id ?? this.id,
      title: title ?? this.title,
      duration: duration ?? this.duration,
      tagsFormat: tagsFormat ?? this.tagsFormat,
      fileType: fileType ?? this.fileType,
      fileSignature: fileSignature ?? this.fileSignature,
      coverSignature: coverSignature ?? this.coverSignature,
      albums: albums ?? this.albums,
      artists: artists ?? this.artists,
      trackNumber: trackNumber ?? this.trackNumber,
      discNumber: discNumber ?? this.discNumber,
      metadata: metadata ?? this.metadata,
      sampleRate: sampleRate ?? this.sampleRate,
      bitRate: bitRate ?? this.bitRate,
      bitsPerRawSample: bitsPerRawSample ?? this.bitsPerRawSample,
      dateAdded: dateAdded ?? this.dateAdded,
      score: score ?? this.score,
      historyInfo: historyInfo != null ? historyInfo() : this.historyInfo,
    );
  }

  @override
  List<Object?> get props => [
        id,
        title,
        duration,
        tagsFormat,
        fileType,
        fileSignature,
        coverSignature,
        albums,
        artists,
        trackNumber,
        discNumber,
        metadata,
        sampleRate,
        bitRate,
        bitsPerRawSample,
        dateAdded,
        score,
        historyInfo,
      ];

  String getUrl(AppSettingAudioQuality quality) {
    switch (quality) {
      case AppSettingAudioQuality.low:
        return "${AppApi().getServerUrl()}track/$id/audio/low/transcode";
      case AppSettingAudioQuality.medium:
        return "${AppApi().getServerUrl()}track/$id/audio/medium/transcode";
      case AppSettingAudioQuality.high:
        return "${AppApi().getServerUrl()}track/$id/audio/high/transcode";
      default:
        return "${AppApi().getServerUrl()}track/$id/audio";
    }
  }

  String getOriginalCoverUrl() {
    return "${AppApi().getServerUrl()}track/$id/cover";
  }

  String getCompressedCoverUrl(TrackCompressedCoverQuality quality) {
    switch (quality) {
      case TrackCompressedCoverQuality.small:
        return "${AppApi().getServerUrl()}track/$id/cover/small";
      case TrackCompressedCoverQuality.medium:
        return "${AppApi().getServerUrl()}track/$id/cover/medium";
      case TrackCompressedCoverQuality.high:
        return "${AppApi().getServerUrl()}track/$id/cover/high";
      default:
        return "${AppApi().getServerUrl()}track/$id/cover";
    }
  }

  Uri getOrignalCoverUri() {
    final url = getOriginalCoverUrl();
    Uri? uri = Uri.tryParse(url);
    if (uri != null && (uri.scheme == 'http' || uri.scheme == 'https')) {
      return uri;
    }
    return Uri.parse("file://$url");
  }

  Uri getCompressedCoverUri(TrackCompressedCoverQuality quality) {
    final url = getCompressedCoverUrl(quality);
    Uri? uri = Uri.tryParse(url);
    if (uri != null && (uri.scheme == 'http' || uri.scheme == 'https')) {
      return uri;
    }
    return Uri.parse("file://$url");
  }

  String getQualityText() {
    if (bitsPerRawSample != null) {
      return "$bitsPerRawSample-Bit ${doubleToStringWithoutZero(sampleRate / 1000)}KHz $fileType";
    }

    if (bitRate != null) {
      return "${doubleToStringWithoutZero(bitRate! / 1000)} kbps $fileType";
    }

    return "Unknown $fileType";
  }
}

class TrackMetadata extends Equatable {
  final int totalTracks;
  final int totalDiscs;

  final String date;
  final int year;

  final List<String> genres;
  final String lyrics;
  final String comment;

  final String acoustId;

  final String musicBrainzReleaseId;
  final String musicBrainzTrackId;
  final String musicBrainzRecordingId;

  final String composer;

  const TrackMetadata({
    required this.totalTracks,
    required this.totalDiscs,
    required this.date,
    required this.year,
    required this.genres,
    required this.lyrics,
    required this.comment,
    required this.acoustId,
    required this.musicBrainzReleaseId,
    required this.musicBrainzTrackId,
    required this.musicBrainzRecordingId,
    required this.composer,
  });

  TrackMetadata copyWith({
    int? totalTracks,
    int? totalDiscs,
    String? date,
    int? year,
    List<String>? genres,
    String? lyrics,
    String? comment,
    String? acoustId,
    String? musicBrainzReleaseId,
    String? musicBrainzTrackId,
    String? musicBrainzRecordingId,
    String? composer,
  }) {
    return TrackMetadata(
      totalTracks: totalTracks ?? this.totalTracks,
      totalDiscs: totalDiscs ?? this.totalDiscs,
      date: date ?? this.date,
      year: year ?? this.year,
      genres: genres ?? this.genres,
      lyrics: lyrics ?? this.lyrics,
      comment: comment ?? this.comment,
      acoustId: acoustId ?? this.acoustId,
      musicBrainzReleaseId: musicBrainzReleaseId ?? this.musicBrainzReleaseId,
      musicBrainzTrackId: musicBrainzTrackId ?? this.musicBrainzTrackId,
      musicBrainzRecordingId:
          musicBrainzRecordingId ?? this.musicBrainzRecordingId,
      composer: composer ?? this.composer,
    );
  }

  @override
  List<Object?> get props => [
        totalTracks,
        totalDiscs,
        date,
        year,
        genres,
        lyrics,
        comment,
        acoustId,
        musicBrainzReleaseId,
        musicBrainzTrackId,
        musicBrainzRecordingId,
        composer,
      ];
}
